Lib/getpass.py

```diff
@@ -186,33 +238,160 @@ def _raw_input(prompt="", stream=None, input=None, echo_char=None):
+            self.cursor_pos -= 1
+        # Delete the word
+        while self.cursor_pos > 0 and self.passwd[self.cursor_pos-1] != ' ':
+            self.cursor_pos -= 1
+        # Remove the deleted portion
+        self.passwd = self.passwd[:self.cursor_pos] + self.passwd[old_cursor:]
+        self.refresh_display()
+        self.eof_pressed = False
+
+    def build_dispatch_table(self):
+        """Build dispatch table mapping control chars to handlers."""
+        return {
+            self.ctrl['SOH']: self.handle_move_start,     # Ctrl+A
+            self.ctrl['ENQ']: self.handle_move_end,       # Ctrl+E
+            self.ctrl['VT']: self.handle_kill_forward,    # Ctrl+K
+            self.ctrl['KILL']: self.handle_kill_line,     # Ctrl+U
+            self.ctrl['WERASE']: self.handle_erase_word,  # Ctrl+W
+            self.ctrl['ERASE']: self.handle_erase,        # DEL
+            '\b': self.handle_erase,                      # Backspace
+        }
+
+
+def _readline_with_echo_char(stream, input, echo_char, term_ctrl_chars=None):
+    """Read password with echo character and line editing support."""
+    if term_ctrl_chars is None:
+        term_ctrl_chars = _POSIX_CTRL_CHARS.copy()
+
+    editor = _PasswordLineEditor(stream, echo_char, term_ctrl_chars)
+    dispatch = editor.build_dispatch_table()
+
     while True:
         char = input.read(1)
-        if char == '\n' or char == '\r':
+
+        # Check for line terminators
+        if char in ('\n', '\r'):
             break
-        elif char == '\x03':
+
+        # Handle literal next mode FIRST (Ctrl+V quotes next char)
+        if editor.literal_next:
+            editor.handle_literal_next(char)
+            continue
+
+        # Check if it's the LNEXT character
+        if char == editor.ctrl['LNEXT']:
+            editor.literal_next = True
+            editor.eof_pressed = False
+            continue
+
+        # Check for special control characters
+        if char == editor.ctrl['INTR']:
             raise KeyboardInterrupt
-        elif char == '\x7f' or char == '\b':
-            if passwd:
-                stream.write("\b \b")
-                stream.flush()
-            passwd = passwd[:-1]
-        elif char == '\x04':
-            if eof_pressed:
+        if char == editor.ctrl['EOF']:
+            if editor.eof_pressed:
                 break
-            else:
-                eof_pressed = True
-        elif char == '\x00':
+            editor.eof_pressed = True
+            continue
+        if char == '\x00':
             continue
+
+        # Dispatch to handler or insert as normal character
+        handler = dispatch.get(char)
+        if handler:
+            handler()
         else:
-            passwd += char
-            stream.write(echo_char)
-            stream.flush()
-            eof_pressed = False
-    return passwd
+            editor.insert_char(char)
+            editor.eof_pressed = False
```

Comment:
Here's a suggestion: don't build a dispatch table via a function but make the editor handles it internally:

```py
editor.handle(char)
```